Update README.md (#147)
schectma/vuln-bank-467_W26: README.md

# Web Security Research — Vulnerable Bank

A web security research project conducted as part of CS 467 (Online Capstone) at Oregon State University, Winter 2026. This project is student managed and directed.

## Disclaimer

**This application is for web security research purposes only.** 

> The Vulnerable Bank web application was not developed by this team. It is an existing open source, intentionally vulnerable banking app. Our work focuses on **demonstrating exploits** against its built-in vulnerabilities, **developing mitigations**, and **documenting** both the attacks and defenses so that anyone can reproduce the results.

Reviewer: ```suggestion
The Vulnerable Bank web application was not developed by this team. It is [an existing open source, intentionally vulnerable banking app](https://github.com/Commando-X/vuln-bank) developed by GitHub user [Commando-X.](https://github.com/Commando-X) Our work focuses on **demonstrating exploits** against its built-in vulnerabilities, **developing mitigations**, and **documenting** both the attacks and defenses so that anyone can reproduce the results.
```

- [Password Hashing Modes](DOCS/hashing.md)
- [Password Cracking Demonstrations](DOCS/hashing_cracking.md)

## Tech Stack

The upstream Vulnerable Bank application uses:

- **Backend:** Python 3.9, Flask 2.0.1, PostgreSQL 13, PyJWT, Argon2
- **Frontend:** HTML, CSS, JavaScript (Jinja2 templates)
- **Deployment:** Docker, Docker Compose, Google Cloud Platform (GCP)
- **Testing:** pytest
- **AI Integration:** [OpenRouter API](https://openrouter.ai/) (NVIDIA Nemotron)

## Live Instance

A hosted instance of the application is available for testing on Google Cloud Platform:

**http://34.28.235.106/**

## Getting Started

### Prerequisites

- [Docker](https://docs.docker.com/get-docker/) and [Docker Compose](https://docs.docker.com/compose/install/) (recommended)
- Or: Python 3.9+ and PostgreSQL 13

### Quick Start with Docker (Recommended)

1. Clone the repository:
   ```bash
   git clone https://github.com/schectma/vuln-bank-467_W26.git
   cd vuln-bank-467_W26
   ```

2. Copy the example environment file and configure it:
   ```bash
   cp .env.example .env
   ```

> 3. (Optional) Add your [OpenRouter](https://openrouter.ai/) API key to `.env` if you want AI chat functionality:

Reviewer: ```suggestion
3. (Optional) Add your [OpenRouter](https://openrouter.ai/) API key to `.env` if desired:
```
We already have a working key so this isn't strictly necessary to enable AI chat functionality. Acknowledged that someone might want to use their own though

   git clone https://github.com/schectma/vuln-bank-467_W26.git
   cd vuln-bank-467_W26
   python3 -m venv venv
   source venv/bin/activate
   pip install -r requirements.txt
   ```

2. Ensure PostgreSQL is running and update `.env` with your database credentials (set `DB_HOST=localhost`).

3. Start the application:
   ```bash
   python3 app.py
   ```

## Usage

### Default Accounts

| Role | Username | Password |
|---|---|---|
| Admin | `admin` | `admin123` |

New users can register through the UI and start with a $1,000 balance.

### Toggling Security Protections

The landing page includes a **Harden** toggle button that enables or disables all security mitigations at once. Individual protections (SQL injection, hashing, CORS, etc.) can also be toggled independently through the API.

- **Hardening off (default):** The application runs in its vulnerable state, all built in weaknesses are active.
- **Hardening on:** Parameterized queries, input validation, security headers, rate limiting, and other protections are activated.

This toggle system allows reviewers to easily test each vulnerability in both states without restarting the application.

### Resetting the Database

> A **Reset Database** button on the landing page wipes all user created data and re-seeds the database with the default accounts and test data from `seed_data.sql`. This is useful after testing exploits that modify or corrupt data.

Reviewer: ```suggestion
A **Reset Database** button on the landing page and in the admin control panel wipes all user created data and re-seeds the database with the default accounts and test data from `seed_data.sql`. This is useful after testing exploits that modify or corrupt data.
```

├── database.py             # PostgreSQL schema, connection pooling, seed data
├── ai_agent_deepseek.py    # AI chatbot integration (OpenRouter API)
├── attack.html             # Standalone CSRF attack demo page
├── templates/              # Jinja2 HTML templates (login, dashboard, admin, etc.)
├── static/                 # CSS, JS, uploaded files
├── mitigations/            # Hardened implementations for each vulnerability
│   ├── sql_injections.py   #   Parameterized queries
│   ├── BOLA.py             #   Authorization checks
│   ├── hashing.py          #   Argon2 password hashing
│   ├── MA.py               #   Field whitelisting
│   ├── SSRF.py             #   URL validation
│   ├── session_exp.py      #   Token expiration
│   ├── rate_limiting.py    #   Request rate enforcement
│   └── AI.py               #   AI prompt guardrails
├── seed_data.sql           # SQL seed data for default accounts and test data
├── exploits/               # Exploit payloads for demos
├── tests/                  # pytest test suite
├── DOCS/                   # Vulnerability write-ups and guides
├── Dockerfile
├── docker-compose.yml
├── requirements.txt
└── .env.example
```

## Testing

> Run the test suite with Docker:

Reviewer: ```suggestion
Build the test container/environment with Docker:
```
(The command below doesn't actually run the suite)

```bash
docker-compose -f docker-compose-test.yml up --build
```

Or locally (requires a running PostgreSQL instance):
```bash
pytest tests/
```

See [DOCS/tests.md](DOCS/tests.md) for details on the test suite.

## Documentation

Each write up in [DOCS/](DOCS/) covers:
- What the vulnerability is and its CWE classification
- Where it exists in the application
- Step by step instructions to exploit it
- How the corresponding mitigation works
- How to verify the mitigation is effective

The documentation is designed so that anyone can reproduce the results.

## Team

CS 467 — Online Capstone, Oregon State University, Winter 2026. This project is student managed and directed.

- [Jessica Johnson](https://github.com/JessJohn0)

> - [Alexander Schetman](https://github.com/schectma)

Reviewer: ```suggestion
- [Alexander Schectman](https://github.com/schectma)
```

> - [Alejandro Rodriguez](https://github.com/alejandrorodvar)

Reviewer 2: ```suggestion
- [Alejandro Rodriguez Varona](https://github.com/alejandrorodvar)
```
